src/vcpkg/tools.cpp

```diff
@@ -41,6 +43,57 @@ namespace
         {"netbsd", ToolOs::NetBsd},
         {"solaris", ToolOs::Solaris},
     };
+
+    template<class Key, class Value, class Compare = std::less<>>
+    struct ContextCache
+    {
+        template<class KeyIsh,
+                 class F,
+                 std::enable_if_t<std::is_constructible_v<Key, const KeyIsh&> &&
+                                      detail::is_callable<Compare&, const Key&, const KeyIsh&>::value,
+                                  int> = 0>
+        const Value* get_lazy(DiagnosticContext& context, const KeyIsh& k, F&& f) const
+        {
+            auto it = m_cache.lower_bound(k);
+            // lower_bound returns the first iterator such that it->first is greater than or equal to than k, so k must
+            // be less than or equal to it->first. If k is greater than or equal it->first, then it must be equal so we
+            // have a cache hit.
+            if (it == m_cache.end() || m_cache.key_comp()(k, it->first))
+            {
+                ContextBufferedDiagnosticContext cbdc{context};
+                auto maybe_result = f(context);
```

Comment:
```suggestion
                ContextBufferedDiagnosticContext cbdc{context};
                auto maybe_result = f(cbdc);
```

AND those new tests found a bug :)